### PR TITLE
[emscripten::val] Support literal utf-8/16 strings

### PR DESCRIPTION
**Pull request by xbcnn**
It already has supports for cstrings with Latin1 encoding. And std::string
could be used for utf-8 encoded strings, but with cost on extra copy of
wire type.
I don't want to extend cstring for utf encoding so that it keeps fast in
most cases. Instead, add new static functions to specially support u8string
and u16string.

### CODE REVIEW

**Review by sbc100 — APPROVED** [review]
Looks good to me .. but my embind knowledge is limited.  WDYT @RReverser ?

File: system/include/emscripten/val.h [hunk 1]
```diff
@@ -348,6 +350,14 @@ namespace emscripten {
             return val(internal::_emval_new_object());
         }
 
+        static val u8string(const char* s) {
```
Comment (RReverser) [inline system/include/emscripten/val.h:353]: Should probably use `char8_t`.

Comment (sbc100) [reply system/include/emscripten/val.h:353]: I think that is C++10 only

Comment (xbcnn) [reply system/include/emscripten/val.h:353]: Yeah. `char8_t` needs C++20 and later. It's a distinct type for UTF-8 strings and has underlying type `unsigned char`.

According to Tom Honermann:
>https://stackoverflow.com/questions/57402464/is-c20-char8-t-the-same-as-our-old-char

Seems `char8_t` is more restrictive and may not accept literal strings without `u8` prefixed.  
We may overload it with `const char8_t*` later.

Comment (RReverser) [reply system/include/emscripten/val.h:353]: Oh ok, weird that they added u8string before the dedicated char type.

**Comment by xbcnn** [comment]
I added the overloading functions for C++20 `char8_t`.  @sbc100 @RReverser, please take a look and give comments.

File: system/include/emscripten/val.h [hunk 2]
```diff
@@ -385,6 +401,17 @@ namespace emscripten {
             : handle(internal::_emval_new_cstring(v))
         {}
 
+#ifndef _LIBCPP_HAS_NO_CHAR8_T
```
Comment (sbc100) [inline system/include/emscripten/val.h:404]: This seems rather odd.. can we not do a positive check?  Maybe just skip this and stick to `const char*` for now ..  unless there is some actual benefit?  (I don't think there is, is there?)

Comment (xbcnn) [reply system/include/emscripten/val.h:404]: Oh, a bit of weird. I just searched the repo and got it from here:
https://github.com/emscripten-core/emscripten/blob/main/system/lib/libcxx/include/__config#L987
> #if _LIBCPP_STD_VER <= 17 || !defined(__cpp_char8_t)
> #define _LIBCPP_HAS_NO_CHAR8_T
> #endif
We can use the standard `__cpp_char8_t` if necessary.  But I agree with you that `const char*` is good for now.  We can do this later when need.  I'll revert the second commit.

**Comment by xbcnn** [comment]
@sbc100 @RReverser  I have rebased the branch to the latest main and seems all tests passed. Do you have any further concerns about the PR?

**Comment by xbcnn** [comment]
@sbc100  Please help to land this.

**Review by kripken — APPROVED** [review]
lgtm. @sbc100 did you have any last comments?

File: system/include/emscripten/val.h [hunk 3]
```diff
@@ -351,6 +353,14 @@ class val {
         return val(internal::_emval_new_object());
     }
 
+    static val u8string(const char* s) {
+        return val(internal::_emval_new_u8string(s));
+    }
+
+    static val u16string(const char16_t* s) {
+        return val(internal::_emval_new_u16string(s));
+    }
```
Comment (kripken) [inline system/include/emscripten/val.h:362]: Oh, I think these should be documented here: https://emscripten.org/docs/api_reference/val.h.html (in the repo, under `site/`).

**Review by sbc100 — APPROVED** [review]

File: system/include/emscripten/val.h (hunk 3 above)
Comment (xbcnn) [reply system/include/emscripten/val.h:362]: OK.

**Pull request merged**